File: mrvtools/ghg_inventory/report/ghg_inventory_report_year_wise/ghg_inventory_report_year_wise.py
# Copyright (c) 2023, tridotstech and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
	columns, data, chart = getColumns(filters),getData(filters), get_chart(filters)
	return columns, data , None, chart


def getColumns(filters):
	columns = []
	columns.append("Categories" + ":Data:500")

	val = f"""
			SELECT 
				name 
			FROM 
				`tabGHG Inventory Master Report` 
			WHERE 
				name >= '{filters.get("from_year")}' 
			AND 
				name <= '{filters.get("to_year")}' 
			ORDER BY 
				year ASC;

		"""
	data = frappe.db.sql(val,as_dict =1)
	frappe.log_error("Datas",data)
	for i in data:
		columns.append(f'{i.name}'+":Float:160")
	return columns

def getData(filters):
	global chart_label
	global co2_value
	global ch4_value
	global n2o_value
	chart_label = []
	co2_value = []
	ch4_value = []
	n2o_value = []
	if not filters.get("inventory_unit") or filters.get("inventory_unit") == 'tCO2e':	
		query = f"""
				Select
					categories
				from
					`tabGHG Inventory Master Report ChildTable`
				where
					parent = "1990"
				order by
					idx
				"""
		data = frappe.db.sql(query,as_dict =1)
		val = f"""
				SELECT 
					name 
				FROM 
					`tabGHG Inventory Master Report` 
				WHERE 
					name >= '{filters.get("from_year")}' 
				AND 
					name <= '{filters.get("to_year")}' 
				ORDER BY 
					year ASC;

			"""
		value = frappe.db.sql(val,as_dict = 1)
		for i in value:
			chart_label.append(i.name)
		
		for each in data:
			for i in value:
				get_total_co2 = f"""
								SELECT 
									total_co2_eq,parent
								FROM 
									`tabGHG Inventory Master Report ChildTable` 
								WHERE 
									parent = '{i.name}' 
								AND
									categories = '{each.categories}'
								ORDER BY 
									idx
							"""
				
				total_co2 = frappe.db.sql(get_total_co2,as_dict =1)

				# frappe.log_error("Total",total_co2)
				each[f'{i.name}'] = total_co2[0].total_co2_eq if total_co2 and total_co2[0].total_co2_eq else 0

				chart_value = f"""
						SELECT 
									co2,ch4,n2o
								FROM 
									`tabGHG Inventory Master Report ChildTable` 
								WHERE 
									parent = '{i.name}' 
								AND
									categories = 'Total National Emissions and Removals'
					"""
				chart_data = frappe.db.sql(chart_value,as_dict =1)
				co2_value.append(chart_data[0].co2)
				ch4_value.append(chart_data[0].ch4)
				n2o_value.append(chart_data[0].n2o)
		
		return data

	if filters.get("inventory_unit") == 'GgCO2e':
		query = f"""
				Select
					categories
				from
					`tabGHG Inventory Master Report ChildTable`
				where
					parent = "1990"
				order by
					idx
				"""
		data = frappe.db.sql(query,as_dict =1)
		val = f"""
				SELECT 
					name 
				FROM 
					`tabGHG Inventory Master Report` 
				WHERE 
					name >= '{filters.get("from_year")}' 
				AND 
					name <= '{filters.get("to_year")}' 
				ORDER BY 
					year ASC;

			"""
		value = frappe.db.sql(val,as_dict = 1)
		for i in value:
			chart_label.append(i.name)
		
		for each in data:
			for i in value:
				get_total_co2 = f"""
								SELECT 
									total_co2_eq * 1000 as total_co2_eq,
									parent
								FROM 
									`tabGHG Inventory Master Report ChildTable` 
								WHERE 
									parent = '{i.name}' 
								AND
									categories = '{each.categories}'
								ORDER BY 
									idx
							"""
				
				total_co2 = frappe.db.sql(get_total_co2,as_dict =1)

				# frappe.log_error("Total",total_co2)
				each[f'{i.name}'] = total_co2[0].total_co2_eq if total_co2 and total_co2[0].total_co2_eq else 0

				chart_value = f"""
						SELECT 
									co2 * 1000 as co2,
									ch4 * 1000 as ch4,
									n2o * 1000 as n2o
								FROM 
									`tabGHG Inventory Master Report ChildTable` 
								WHERE 
									parent = '{i.name}' 
								AND
									categories = 'Total National Emissions and Removals'
					"""
				chart_data = frappe.db.sql(chart_value,as_dict =1)
				co2_value.append(chart_data[0].co2)
				ch4_value.append(chart_data[0].ch4)
				n2o_value.append(chart_data[0].n2o)
		
		return data
def get_chart(filters):
	
	val = f"""
			SELECT 
				name 
			FROM 
				`tabGHG Inventory Master Report` 
			WHERE 
				name >= '{filters.get("from_year")}' 
			AND 
				name <= '{filters.get("to_year")}' 
			ORDER BY 
				year ASC;

		"""
	value = frappe.db.sql(val,as_dict = 1)

	

	chart = {
		"data": {
			"labels": chart_label,
			"datasets": [
				{
					"name": "CO2",
					"values": co2_value,
					"chartType": "bar",
				},
				{
					"name": "CH4",
					"values": ch4_value,
					"chartType": "bar"
				},
				{
					"name": "N2O",
					"values": n2o_value,
					"chartType": "bar"
				}
			]
		},
		"barOptions": {
			"stacked": 1
		},
		"colors":["pink","blue","green"]
	}
	return chart

File: mrvtools/ghg_inventory/report/ghg_inventory_report_year_wise/ghg_inventory_report_year_wise.js
// Copyright (c) 2023, tridotstech and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["GHG Inventory Report Year Wise"] = {
	"filters": [
		{
			"fieldname": "from_year",
			"fieldtype": "Select",
			"label": __("From Year"),
			"options": [" ","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017","2018","2019","2020","2021","2022","2023","2024","2025","2026","2027","2028","2029","2030","2031","2032","2033","2034","2035","2036","2037","2038","2039","2040","2041","2042","2043","2044","2045","2046","2047","2048","2049","2050"],
			change:function(){
				var data = $('[data-route="query-report/GHG Inventory Report Year Wise"] select[data-fieldname="from_year"]').val()
				frappe.query_report.filters[0].value = data

				frappe.query_report.filters[0].placeholder = data ? "" : "From Year"
				frappe.query_report.filters[0].refresh()
				// if (data != ""){
				// 	frappe.query_report.filters[0].placeholder = ""
				// 	frappe.query_report.filters[0].refresh()
				// }
				// else{
				// 	frappe.query_report.filters[0].placeholder = "From Year"
				// 	frappe.query_report.filters[0].refresh()
				// }
					
				var option = []
				for(let i = parseInt(data);i<=(parseInt(data)+10);i++){
					option.push(i)
				}

				
				var year_filter = frappe.query_report.get_filter('to_year');
				year_filter.df.options = option;
				// year_filter.df.default = r.message.split("\n")[0];
				year_filter.refresh();
				year_filter.set_input([""]);
				
			}
		},
		{
			"fieldname": "to_year",
			"fieldtype": "Select",
			"label": __("To Year"),	
			"options": [""]
		},
		{
			"fieldname": "inventory_unit",
			"fieldtype": "Select",
			"label": __("Inventory Unit"),
			"options": [" ","tCO2e","GgCO2e"]
		}
	]
};
